File: citizenmate/scripts/generate-images.py
#!/usr/bin/env python3
"""
CitizenMate - Kie.ai Nano Banana 2 Image Generator
Generates ultra-realistic, consistent images for the landing page.

Usage: python3 scripts/generate-images.py
"""

import json
import os
import ssl
import sys
import time
import urllib.request
import urllib.error

# Fix macOS SSL certificate issue
ssl_ctx = ssl.create_default_context()
try:
    import certifi
    ssl_ctx.load_verify_locations(certifi.where())
except ImportError:
    ssl_ctx.check_hostname = False
    ssl_ctx.verify_mode = ssl.CERT_NONE

API_KEY = os.environ.get("KIE_API_KEY", "e4fc401761f2af571ada5f8837de3c63")
API_BASE = "https://api.kie.ai/api/v1/jobs"
OUTPUT_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "public", "generated")
MAX_POLL_SECONDS = 180
POLL_INTERVAL = 5

# Ultra-realistic style directive
STYLE = (
    "Ultra-realistic, DSLR-quality editorial photograph shot on Canon EOS R5 "
    "with 85mm f/1.4 lens. Natural skin with visible pores, fine hair, and "
    "real texture. Absolutely no airbrushed, plastic, or AI-generated skin "
    "appearance. Soft natural golden-hour lighting. Subtle film grain for "
    "authenticity. Shallow depth of field. Color palette: deep navy blue, "
    "warm gold, eucalyptus green, sky blue. Stock photography quality, "
    "indistinguishable from a real professional photo."
)

IMAGES = [
    {
        "filename": "hero-study.png",
        "aspect_ratio": "16:9",
        "resolution": "2K",
        "prompt": (
            "A warm, inviting scene of a diverse multicultural group of adults "
            "studying together at a modern Australian library. The group includes "
            "people of South Asian, East Asian, Middle Eastern, African, and Latin "
            "American backgrounds, all smiling, engaged, and helping each other learn. "
            "Visible through large windows: iconic Australian harbour, blue sky, "
            "eucalyptus trees. Open books, a tablet showing quiz questions, and "
            "coffee cups on the table. The mood is hopeful, supportive, and "
            "celebratory of new beginnings in Australia. Natural candid moment."
        ),
    },
    {
        "filename": "feature-tests.png",
        "aspect_ratio": "3:2",
        "resolution": "2K",
        "prompt": (
            "Close-up of a confident young woman of South Asian descent using "
            "a modern tablet to take a practice citizenship test. The screen "
            "shows a clean quiz interface with multiple-choice questions. She is "
            "seated at a bright airy desk with a small Australian flag pin on her "
            "notebook. Soft bokeh background of a modern Australian home interior. "
            "Expression: focused yet calm and empowered. Natural candid portrait."
        ),
    },
    {
        "filename": "feature-bilingual.png",
        "aspect_ratio": "3:2",
        "resolution": "2K",
        "prompt": (
            "A warm scene showing a young man of East Asian background studying "
            "with a bilingual book, English on one page and Chinese characters on "
            "the facing page. He is seated in a cozy Australian cafe with eucalyptus "
            "sprigs in a vase nearby. Books in multiple languages stacked beside him: "
            "Arabic, Hindi, Vietnamese, Mandarin. The mood is inclusive and "
            "empowering, learning in your own language to succeed in English. "
            "Natural warm candid portrait."
        ),
    },
    {
        "filename": "feature-progress.png",
        "aspect_ratio": "3:2",
        "resolution": "2K",
        "prompt": (
            "A joyful young woman of African background looking at a laptop screen "
            "showing a dashboard with progress charts, green checkmarks, and "
            "achievement badges. She is smiling with a fist pump of celebration. "
            "Setting: bright modern Australian apartment with sunny suburban view "
            "through the window. The laptop screen shows circular progress rings at "
            "85 percent completion in green and gold colors. Mood: achievement, "
            "confidence, readiness. Candid natural shot."
        ),
    },
    {
        "filename": "how-it-works-bg.png",
        "aspect_ratio": "16:9",
        "resolution": "2K",
        "prompt": (
            "A serene, wide Australian landscape background. Soft rolling "
            "eucalyptus-covered hills under a vast blue sky with scattered white "
            "clouds. Golden afternoon sunlight bathes the scene. Gentle peaceful "
            "color palette of sage green, sky blue, soft gold, and cream. Clean "
            "minimal composition. Slightly dreamy atmosphere with a sense of "
            "openness, welcome, and possibility. No people. Real landscape "
            "photography."
        ),
    },
    {
        "filename": "social-proof-people.png",
        "aspect_ratio": "16:9",
        "resolution": "2K",
        "prompt": (
            "A heartwarming group photo of newly inducted Australian citizens at "
            "an outdoor citizenship ceremony. A diverse group of 8 adults from "
            "different cultures, wearing smart casual clothes, some holding small "
            "Australian flags, all beaming with pride and joy. Behind them is a "
            "large Australian coat of arms banner. The group includes people of "
            "Indian, Filipino, Chinese, Lebanese, Somali, Colombian, and Vietnamese "
            "backgrounds. Children present too. The mood is emotional, proud, and "
            "joyously celebratory. Natural group photograph."
        ),
    },
    {
        "filename": "cta-australia.png",
        "aspect_ratio": "16:9",
        "resolution": "2K",
        "prompt": (
            "Breathtaking panoramic view of Sydney Harbour at golden hour. The "
            "Opera House and Harbour Bridge illuminated in warm golden light "
            "reflecting off calm water. Silhouettes of a small diverse family, "
            "parents of different cultural backgrounds with two children, standing "
            "on a lookout point at the railing, gazing at the view with arms around "
            "each other. The sky painted in warm oranges, golds, and soft blues. "
            "Mood: hopeful, forward-looking, a family beginning an exciting new "
            "chapter in Australia. Cinematic landscape photography."
        ),
    },
    {
        "filename": "og-image.png",
        "aspect_ratio": "16:9",
        "resolution": "2K",
        "prompt": (
            "A professional, clean hero composition for a social media card. On the "
            "left side: a warm photo of a diverse group of 5 happy multicultural "
            "adults holding small Australian flags at a citizenship ceremony, dressed "
            "smartly. On the right side: a clean modern phone showing a quiz app with "
            "green checkmarks. Background: soft elegant gradient from deep navy blue "
            "to warm gold. Subtle Australian Southern Cross star pattern. "
            "Professional, trustworthy, and inviting composition."
        ),
    },
    {
        "filename": "dash-welcome.png",
        "aspect_ratio": "21:9",
        "resolution": "2K",
        "prompt": (
            "A welcoming over-the-shoulder shot of a diverse individual sitting at "
            "a clean, modern desk in an Australian home, looking at a beautiful "
            "tablet displaying a study plan. A stylized map of Australia is pinned "
            "to the corkboard in front of them. A steaming coffee cup and eucalyptus "
            "leaves on the desk. Warm, bright morning sunlight streaming through "
            "the window, representing a fresh start and motivation for the day's study."
        ),
    },
    {
        "filename": "dash-goal.png",
        "aspect_ratio": "3:2",
        "resolution": "2K",
        "prompt": (
            "Close-up, top-down view of multicultural hands using a nice pen to "
            "confidently circle a date on a physical desktop calendar. Next to the "
            "calendar sits a small Australian flag and a sprig of eucalyptus. "
            "Warm golden hour light highlights the hands. Motif of setting a goal "
            "and taking action. Professional, clean, and motivating photography."
        ),
    },
    {
        "filename": "study-header.png",
        "aspect_ratio": "21:9",
        "resolution": "2K",
        "prompt": (
            "An optimistic, wide cinematic shot of a diverse group of four people "
            "sitting casually on the grass in a sunny, beautiful Australian park. "
            "They are laughing and discussing a study booklet ('Our Common Bond' style), "
            "sharing knowledge. In the background, softly blurred Australian flora "
            "and a hint of a city skyline. Symbolizes the shared journey of learning "
            "Australian history and values together."
        ),
    },
    {
        "filename": "topic-people.png",
        "aspect_ratio": "3:2",
        "resolution": "2K",
        "prompt": (
            "A beautiful editorial portrait capturing the diversity of Australia. "
            "A group of three modern Australians (Indigenous, Anglo-Celtic, and "
            "East Asian heritage) sharing a laugh over a casual outdoor meal. "
            "Subtle, tasteful Aboriginal art motifs seamlessly integrated into the "
            "modern cafe background wall. Warm, inviting, and representative of "
            "a harmonious, multicultural society."
        ),
    },
    {
        "filename": "topic-beliefs.png",
        "aspect_ratio": "3:2",
        "resolution": "2K",
        "prompt": (
            "Breathtaking twilight photography of the Australian Parliament House "
            "in Canberra. The distinctive flagpole is illuminated against a deep "
            "navy and magenta sunset sky. Clean, majestic, and awe-inspiring. "
            "Symbolizes democracy, freedom of speech, and the core rights and "
            "liberties of Australian citizens. Shot from a low, empowering angle."
        ),
    },
    {
        "filename": "topic-gov.png",
        "aspect_ratio": "3:2",
        "resolution": "2K",
        "prompt": (
            "A professional, modern scene inside a bright, contemporary Australian "
            "government building or courtroom. Soft, diffused lighting illuminating "
            "wooden textures and the Australian coat of arms subtly in the background. "
            "A diverse professional woman in a crisp suit confidently reviewing a "
            "document. Conveys fairness, equality, governance, and the rule of law "
            "in a modern democratic society."
        ),
    },
    {
        "filename": "test-start.png",
        "aspect_ratio": "3:2",
        "resolution": "2K",
        "prompt": (
            "A close-up portrait of a young man of Middle Eastern descent taking "
            "a deep breath, eyes closed with a gentle, confident smile, holding "
            "a tablet in his hands right before starting a test. Soft, focused, "
            "warm indoor lighting representing deep concentration and mental "
            "preparation. The background is a dark, quiet library setting. "
            "Conveys encouragement and calm readiness."
        ),
    },
    {
        "filename": "result-pass.png",
        "aspect_ratio": "16:9",
        "resolution": "2K",
        "prompt": (
            "Exuberant celebration! A young woman of dynamic mixed heritage "
            "throwing her hands up in absolute joy, cheering loudly. She is holding "
            "up a digital tablet displaying large green checkmarks and a 100% score. "
            "Out of focus in the background, friends are clapping and celebrating "
            "with her. Bright, saturated, incredibly happy colors. Confetti-like "
            "bokeh in the sunlight. Pure elation and relief."
        ),
    },
    {
        "filename": "result-tryagain.png",
        "aspect_ratio": "16:9",
        "resolution": "2K",
        "prompt": (
            "A highly empathetic and supportive scene. A person looking determined "
            "but thoughtful, pointing at a study guide on a table. Sitting right "
            "beside them is an encouraging mentor or friend with a warm, reassuring "
            "hand on their shoulder, smiling supportively. Warm, soft lighting that "
            "feels safe and comforting. The visual message is 'Don't give up, "
            "you will get it next time.' Natural, candid emotion."
        ),
    },
    {
        "filename": "modal-test-date.png",
        "aspect_ratio": "16:9",
        "resolution": "2K",
        "prompt": (
            "A breathtaking, highly stylized sunrise over a famous Australian "
            "beach (like Bondi or Whitehaven). The sky is exploding with vibrant "
            "pinks, oranges, and golds, reflecting perfectly on the gentle surf. "
            "In the foreground, a subtle, beautiful abstract representation of a "
            "calendar page fluttering in the sea breeze. Symbolizes a new dawn, "
            "setting a life-changing goal, and looking forward to the future."
        ),
    },
    {
        "filename": "avatar-tutor.png",
        "aspect_ratio": "1:1",
        "resolution": "2K",
        "prompt": (
            "A hyper-realistic close-up portrait avatar for an AI chatbot. "
            "A remarkably friendly, approachable professional of mixed cultural "
            "heritage (e.g., Afro-Latina or Eurasian), looking directly into the "
            "lens with a genuinely warm, helpful smile. She is wearing a neat, "
            "modern navy-blue jacket. Crisp lighting against a clean, softly blurred "
            "background. Inspires immense trust and approachability."
        ),
    },
]


def api_request(url, method="GET", data=None):
    """Make an API request to kie.ai."""
    headers = {
        "Authorization": f"Bearer {API_KEY}",
        "Content-Type": "application/json",
    }
    body = json.dumps(data).encode() if data else None
    req = urllib.request.Request(url, data=body, headers=headers, method=method)

    try:
        with urllib.request.urlopen(req, timeout=30, context=ssl_ctx) as resp:
            return json.loads(resp.read().decode())
    except urllib.error.HTTPError as e:
        body_text = e.read().decode() if e.fp else ""
        print(f"\n    HTTP {e.code}: {body_text}")
        raise
    except Exception as e:
        print(f"\n    Request error: {e}")
        raise


def create_task(image_def):
    """Create an image generation task."""
    full_prompt = f"{STYLE} {image_def['prompt']}"
    payload = {
        "model": "nano-banana-2",
        "input": {
            "prompt": full_prompt,
            "image_input": [],
            "aspect_ratio": image_def["aspect_ratio"],
            "resolution": image_def["resolution"],
            "output_format": "png",
        },
    }
    result = api_request(f"{API_BASE}/createTask", method="POST", data=payload)
    if result.get("code") != 200:
        raise Exception(f"API error: {result.get('msg', 'unknown')}")
    return result["data"]["taskId"]


def poll_task(task_id):
    """Poll task until success or failure."""
    elapsed = 0
    while elapsed < MAX_POLL_SECONDS:
        result = api_request(f"{API_BASE}/recordInfo?taskId={task_id}")
        state = result.get("data", {}).get("state", "unknown")

        if state == "success":
            result_json = json.loads(result["data"].get("resultJson", "{}"))
            urls = result_json.get("resultUrls", [])
            return urls[0] if urls else None

        if state == "fail":
            msg = result["data"].get("failMsg", "unknown")
            raise Exception(f"Generation failed: {msg}")

        sys.stdout.write(".")
        sys.stdout.flush()
        time.sleep(POLL_INTERVAL)
        elapsed += POLL_INTERVAL

    raise Exception(f"Timed out after {MAX_POLL_SECONDS}s")


def download_file(url, dest_path):
    """Download a file from URL."""
    req = urllib.request.Request(url, headers={
        "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) CitizenMate/1.0",
    })
    with urllib.request.urlopen(req, context=ssl_ctx) as resp:
        with open(dest_path, 'wb') as f:
            f.write(resp.read())


def main():
    print("🎨 CitizenMate — Kie.ai Nano Banana 2 Image Generator\n")
    print(f"📁 Output: {OUTPUT_DIR}")
    print()

    os.makedirs(OUTPUT_DIR, exist_ok=True)

    existing = set(os.listdir(OUTPUT_DIR))
    to_generate = []
    for img in IMAGES:
        if img["filename"] in existing:
            print(f"  ⏭  {img['filename']} — already exists, skipping")
        else:
            to_generate.append(img)

    if not to_generate:
        print("\n✅ All images already generated! Delete files to regenerate.")
        return

    print(f"\n🖼  Generating {len(to_generate)} images...\n")

    success = 0
    failed = 0

    for img in to_generate:
        try:
            sys.stdout.write(f"  🔄 {img['filename']} ")
            sys.stdout.flush()

            task_id = create_task(img)
            sys.stdout.write(f"[{task_id[:12]}...] ")
            sys.stdout.flush()

            image_url = poll_task(task_id)
            if not image_url:
                raise Exception("No result URL returned")

            dest = os.path.join(OUTPUT_DIR, img["filename"])
            download_file(image_url, dest)

            size_kb = os.path.getsize(dest) // 1024
            print(f" ✅ ({size_kb} KB)")
            success += 1

        except Exception as e:
            print(f" ❌ {e}")
            failed += 1

        time.sleep(2)  # Brief delay between requests

    skipped = len(IMAGES) - len(to_generate)
    print(f"\n📊 Results: {success} generated, {failed} failed, {skipped} skipped")
    print("🎉 Done!")


if __name__ == "__main__":
    main()

File: citizenmate/scripts/generate-images.ts
/**
 * Kie.ai Image Generation Script
 * 
 * Pre-generates all images for the CitizenMate landing page using
 * the Nano Banana 2 model via kie.ai API.
 * 
 * Usage: npx tsx scripts/generate-images.ts
 */

import * as fs from "fs";
import * as path from "path";
import * as https from "https";
import * as http from "http";

// ─── Configuration ──────────────────────────────────────────────
const KIE_API_BASE = "https://api.kie.ai/api/v1/jobs";
const API_KEY = process.env.KIE_API_KEY || "e4fc401761f2af571ada5f8837de3c63";
const OUTPUT_DIR = path.resolve(__dirname, "../public/generated");
const MAX_POLL_TIME_MS = 120_000;
const POLL_INTERVAL_MS = 3_000;
const DELAY_BETWEEN_REQUESTS_MS = 2_000; // avoid rate limiting

// ─── Consistent style prefix for all prompts ────────────────────
const STYLE_PREFIX =
  "Ultra-realistic, DSLR-quality editorial photograph shot on Canon EOS R5, 85mm f/1.4 lens. " +
  "Natural skin with visible pores, fine hair, and real texture — absolutely NO airbrushed, plastic, or AI-generated skin appearance. " +
  "Soft natural lighting with golden-hour warmth. Subtle film grain for authenticity. " +
  "Clean modern composition with shallow depth of field. Color palette: deep navy blue, warm gold, eucalyptus green, and sky blue accents. " +
  "Welcoming and aspirational mood. No text overlays unless specified. " +
  "Stock photography quality — indistinguishable from a real professional photo shoot.";

// ─── Image definitions ──────────────────────────────────────────
interface ImageDef {
  filename: string;
  prompt: string;
  aspect_ratio: string;
  resolution: string;
}

const IMAGES: ImageDef[] = [
  {
    filename: "hero-study.png",
    prompt: `${STYLE_PREFIX} A warm, inviting scene of a diverse multicultural group of adults studying together at a modern library or co-working space in Australia. The group includes people of South Asian, East Asian, Middle Eastern, African, and Latin American backgrounds — all smiling and engaged, helping each other learn. Visible through large windows behind them: iconic Australian scenery — harbour, blue sky, gum trees. Open books, a tablet showing quiz questions, and coffee cups on the table. The mood is hopeful, supportive, and celebratory of new beginnings in Australia.`,
    aspect_ratio: "16:9",
    resolution: "2K",
  },
  {
    filename: "feature-tests.png",
    prompt: `${STYLE_PREFIX} Close-up of a confident young woman of South Asian descent using a modern tablet to take a practice citizenship test. The screen shows a clean quiz interface with multiple-choice questions. She is seated at a bright, airy desk with an Australian flag pin on her notebook. Soft bokeh background of a modern Australian suburban home. Expression: focused yet calm. Feeling of preparation and empowerment.`,
    aspect_ratio: "3:2",
    resolution: "2K",
  },
  {
    filename: "feature-bilingual.png",
    prompt: `${STYLE_PREFIX} A warm scene of a young man of East Asian background studying with a bilingual book — English on one page and Chinese characters on the facing page. He is seated in a cozy Australian café with eucalyptus sprigs in a vase nearby. A phone shows a translation app. Books in multiple languages are stacked beside him — Arabic, Hindi, Vietnamese, Mandarin. The mood is inclusive and empowering — learning in your own language to succeed in English.`,
    aspect_ratio: "3:2",
    resolution: "2K",
  },
  {
    filename: "feature-progress.png",
    prompt: `${STYLE_PREFIX} A joyful young woman of African background looking at a laptop screen showing a beautiful dashboard with progress charts, green checkmarks, and achievement badges. She is pumping her fist in celebration. Setting: bright modern Australian apartment with a window showing a sunny suburban view. The laptop screen shows circular progress rings at 85% completion with green and gold colors. Mood: achievement, confidence, readiness.`,
    aspect_ratio: "3:2",
    resolution: "2K",
  },
  {
    filename: "how-it-works-bg.png",
    prompt: `${STYLE_PREFIX} A serene, abstract Australian landscape background. Soft rolling eucalyptus-covered hills under a vast blue sky with scattered clouds. Golden afternoon light bathes the scene. Gentle color palette of sage green, sky blue, soft gold, and cream. Minimal and clean — suitable as a subtle background image. Slightly blurred and dreamy, with a sense of openness and welcome.`,
    aspect_ratio: "16:9",
    resolution: "2K",
  },
  {
    filename: "social-proof-people.png",
    prompt: `${STYLE_PREFIX} A heartwarming group photo of newly inducted Australian citizens at a citizenship ceremony. A diverse group of 8-10 adults from many cultures — wearing smart casual clothes, some holding small Australian flags, all beaming with pride and joy. Behind them is a large Australian coat of arms banner. An older official in a suit is congratulating them. The group includes people of Indian, Filipino, Chinese, Lebanese, Somali, Colombian, and Vietnamese backgrounds. The mood is emotional, proud, and celebratory — the culmination of their citizenship journey.`,
    aspect_ratio: "16:9",
    resolution: "2K",
  },
  {
    filename: "cta-australia.png",
    prompt: `${STYLE_PREFIX} Breathtaking panoramic view of the Sydney Harbour at golden hour. The Opera House and Harbour Bridge are visible in warm golden light reflecting off the water. Silhouettes of a small diverse group of people — a family with children of mixed cultural backgrounds — standing on a lookout point, gazing at the view with arms around each other. The sky is painted in warm oranges, golds, and soft blues. The mood is hopeful, forward-looking — a family's exciting new chapter in Australia.`,
    aspect_ratio: "16:9",
    resolution: "2K",
  },
  {
    filename: "og-image.png",
    prompt: `${STYLE_PREFIX} A professional Open Graph social media card. A warm collage-style composition: on the left, a diverse group of happy multicultural adults holding Australian flags at a citizenship ceremony. On the right, a clean modern phone showing a quiz app with green checkmarks. In the center, bold clean text: "CitizenMate" in navy blue. Background: soft gradient from navy blue to warm gold. Australian Southern Cross stars subtly visible. The mood is professional, trustworthy, and inviting.`,
    aspect_ratio: "16:9",
    resolution: "2K",
  },
];

// ─── API functions ──────────────────────────────────────────────

async function createTask(imageDef: ImageDef): Promise<string> {
  const res = await fetch(`${KIE_API_BASE}/createTask`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${API_KEY}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      model: "nano-banana-2",
      input: {
        prompt: imageDef.prompt,
        image_input: [],
        aspect_ratio: imageDef.aspect_ratio,
        resolution: imageDef.resolution,
        output_format: "png",
      },
    }),
  });

  if (!res.ok) {
    const text = await res.text();
    throw new Error(`createTask failed: ${res.status} — ${text}`);
  }

  const json = await res.json();
  if (json.code !== 200) throw new Error(`createTask API error: ${json.msg}`);
  return json.data.taskId;
}

async function pollTask(taskId: string): Promise<string[]> {
  const start = Date.now();

  while (Date.now() - start < MAX_POLL_TIME_MS) {
    const res = await fetch(`${KIE_API_BASE}/recordInfo?taskId=${taskId}`, {
      headers: { Authorization: `Bearer ${API_KEY}` },
    });

    if (!res.ok) {
      const text = await res.text();
      throw new Error(`pollTask failed: ${res.status} — ${text}`);
    }

    const json = await res.json();
    const state = json.data?.state;

    if (state === "success") {
      const result = JSON.parse(json.data.resultJson || "{}");
      return result.resultUrls || [];
    }

    if (state === "fail") {
      throw new Error(`Generation failed: ${json.data.failMsg || "unknown"}`);
    }

    process.stdout.write(".");
    await sleep(POLL_INTERVAL_MS);
  }

  throw new Error(`Timed out after ${MAX_POLL_TIME_MS / 1000}s`);
}

function downloadFile(url: string, destPath: string): Promise<void> {
  return new Promise((resolve, reject) => {
    const protocol = url.startsWith("https") ? https : http;
    const file = fs.createWriteStream(destPath);

    protocol.get(url, (response) => {
      // Follow redirects
      if (response.statusCode && response.statusCode >= 300 && response.statusCode < 400 && response.headers.location) {
        file.close();
        fs.unlinkSync(destPath);
        downloadFile(response.headers.location, destPath).then(resolve).catch(reject);
        return;
      }

      if (response.statusCode && response.statusCode !== 200) {
        file.close();
        fs.unlinkSync(destPath);
        reject(new Error(`Download failed: HTTP ${response.statusCode}`));
        return;
      }

      response.pipe(file);
      file.on("finish", () => {
        file.close();
        resolve();
      });
    }).on("error", (err) => {
      file.close();
      fs.unlinkSync(destPath);
      reject(err);
    });
  });
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// ─── Main ───────────────────────────────────────────────────────

async function main() {
  console.log("🎨 CitizenMate Image Generator — Kie.ai Nano Banana 2\n");
  console.log(`📁 Output directory: ${OUTPUT_DIR}`);

  // Ensure output dir exists
  fs.mkdirSync(OUTPUT_DIR, { recursive: true });

  // Check which images already exist
  const existingFiles = fs.readdirSync(OUTPUT_DIR);
  const toGenerate = IMAGES.filter((img) => {
    if (existingFiles.includes(img.filename)) {
      console.log(`  ⏭  ${img.filename} — already exists, skipping`);
      return false;
    }
    return true;
  });

  if (toGenerate.length === 0) {
    console.log("\n✅ All images already generated! Delete files to regenerate.");
    return;
  }

  console.log(`\n🖼  Generating ${toGenerate.length} images...\n`);

  let success = 0;
  let failed = 0;

  for (const img of toGenerate) {
    try {
      process.stdout.write(`  🔄 ${img.filename}`);

      // Create task
      const taskId = await createTask(img);
      process.stdout.write(` [task: ${taskId}] `);

      // Poll until done
      const urls = await pollTask(taskId);

      if (!urls.length) {
        throw new Error("No result URLs returned");
      }

      // Download first result
      const destPath = path.join(OUTPUT_DIR, img.filename);
      await downloadFile(urls[0], destPath);

      const stats = fs.statSync(destPath);
      console.log(` ✅ (${(stats.size / 1024).toFixed(0)} KB)`);
      success++;
    } catch (err) {
      console.log(` ❌ ${err instanceof Error ? err.message : err}`);
      failed++;
    }

    // Delay between requests to avoid rate limiting
    if (toGenerate.indexOf(img) < toGenerate.length - 1) {
      await sleep(DELAY_BETWEEN_REQUESTS_MS);
    }
  }

  console.log(`\n📊 Results: ${success} generated, ${failed} failed, ${IMAGES.length - toGenerate.length} skipped`);
  console.log("🎉 Done!\n");
}

main().catch((err) => {
  console.error("Fatal error:", err);
  process.exit(1);
});
